dfusion_rust_core/src/models/account_state.rs

```diff
@@ -14,48 +15,85 @@ use super::util::*;
 pub struct AccountState {
     pub state_hash: H256,
     pub state_index: U256,
-    balances: Vec<u128>,
+    balances: HashMap<U256, HashMap<u8, u128>>, // UserId => (TokenId => balance)
     pub num_tokens: u8,
 }
 
 impl AccountState {
     pub fn new(state_hash: H256, state_index: U256, balances: Vec<u128>, num_tokens: u8) -> Self {
+        assert_eq!(
+            balances.len() % (num_tokens as usize),
+            0,
+            "Elements in balance vector needs to be a multiple of num_tokens"
+        );
         AccountState {
             state_hash,
             state_index,
-            balances,
+            balances: balances
+                .chunks(num_tokens as usize)
+                .enumerate()
+                .map(|(account, token_balances)| {
+                    (
+                        U256::from(account),
+                        token_balances
+                            .iter()
+                            .enumerate()
+                            .map(|(token, balance)| (token as u8, *balance))
+                            .collect(),
+                    )
+                })
+                .collect(),
             num_tokens,
         }
     }
-    fn balance_index(&self, token_id: u8, account_id: u16) -> usize {
-        self.num_tokens as usize * account_id as usize + token_id as usize
+
+    /// Returns the balances as a vector, under the assumptions
+    /// that self.balances contains consecutive addresses (0..num_accounts)
+    /// as keys, and each address has a balance for all (0..num_tokens) tokens.
+    ///
+    /// Panics, if this is not the case.
```

Comment:
nit: I think the panics have a special syntax

https://doc.rust-lang.org/1.30.0/book/first-edition/documentation.html#special-sections

Reply:
# Panics above the panic sentence, good call.